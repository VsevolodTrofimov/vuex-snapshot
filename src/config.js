/**
 * @namespace 
 * @property {Boolean} autoResolve resolve all MockPromises and fetches in order they were created
 * @property {Boolean} snapEnv include state, getters and paylaod into snapshot
 */
const options = {
  autoResolve: false,
  snapEnv: false
}

// they are likely to stay flat
const defaults = Object.assign({}, options)
const reset = () => Object.assign(options, defaults)
reset()

export default {
  options,
  reset
}