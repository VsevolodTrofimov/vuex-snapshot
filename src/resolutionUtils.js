import serialize from './serialize'


export const normalizeResolution = resolution => {
  const normalResolution = {
    name: '',
    type: 'resolve',
    payload: undefined,
  }
  
  // string constructor
  if(typeof resolution === 'string' || resolution.indexOf('string')) {
    normalResolution.name = resolution
    return normalResolution
  } 
  
  // object constructor, errors are duplicated because they are likely to occur in promises
  if(resolution.name) {
    normalResolution.name = resolution.name
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution must have a name')
  }

  if(resolution.type && ['resolve', 'reject'].indexOf(resolution.type) !== -1) {
    normalResolution.type = resolution.type
  } else {
    throw new Error('vuex-snapshot: INPUT ERROR resolution type must be' 
                    + 'either "resovle" or "reject"')
  }

  if(resolution.payload) {
    normalResolution.payload = resolution.payload
  }
}


export const simualteResolution = (resolution, snapCb, timetable) => {
  snapCb(`RESOLUTION: ${resolution.name} -> ${resolution.type}`, resolution.payload)
  return timetable.trigger(resolution)
}


export const simualteResolutions = (resolutions, snapCb, timetable) => {
  return new Promise((resolveSimulation, rejectSimulation) => {
    const normalResolutions = resolutions.map(normalizeResolution)
    
    // simulates given resolution and queues the next until all are simulated
    const simulationLoop = idx => {
      if(idx === normalResolutions.length) {
        resolveSimulation()
      }
      else {
        simualteResolution(normalResolutions[idx], snapCb, timetable)
          .then(() => simulationLoop(idx + 1))
          .catch(rejectSimulation)
      }
    }

    simulationLoop(0)
  })
}
