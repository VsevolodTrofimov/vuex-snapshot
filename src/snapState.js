import serialize from './serialize'


/**
 * Serializes your state the fancy way 
 * (including functions and mock promises)
 * @param {*} state 
 */
export const snapState = state => serialize(state)
export default snapState