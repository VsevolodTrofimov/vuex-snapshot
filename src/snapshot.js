class Snapshot {
  constructor() {
    this.value = []
    this.add = this.add.bind(this)
  }
  add(message, payload) {
    const entry = {}
    entry.message = message
    if(typeof payload !== 'undefined') entry.payload = payload
    this.value.push(entry)
  }
}


export default Snapshot