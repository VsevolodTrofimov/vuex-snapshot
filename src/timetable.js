import {makeSuffix, find} from './utility'


const RealPromise = Promise
export const entries = []


export const register = ({name, promise, payload, resolve, reject}) => {
  const entry = {
    payload,
    promise,
    resolve,
    reject,
    called: false
  }

  // make sure name is unique
  const suffix = makeSuffix()
  while(find(entries, e => e.name === name + suffix())) suffix.next()

  entry.name = name + suffix()

  entries.push(entry)
}


export const trigger = ({name, type, payload}) => {
  const suffix = makeSuffix()
  while(find(entries, e => e.name === name + suffix() && e.called)) suffix.next()
  
  const entry = find(entries, e => e.name === name + suffix())
  
  return new RealPromise((resolve, reject) => {
    if(typeof entry === 'undefined') {
      reject(new Error(`vuex-snapshot: did not find ${name + suffix()} that wasn't already resolved or rejected`))
    }
  
    entry[type](payload)
    entry.called = true
    entry.promise
      .then(resolve)
      .catch(resolve)
  })
}


export const ensureAbsence = (promise) => {
  for(let i = 0; i < entries.length; ++i) {
    if(entries[i].promise === promise) {
      entries.splice(i, 1)
    }
  }
}

export const reset = () => entries.length = 0

export default {
  register,
  trigger,
  reset,
  entries,
  ensureAbsence,
}