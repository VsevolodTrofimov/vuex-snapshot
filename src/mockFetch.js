import timetable from './timetable'
import useGlobally from './useGlobally'


const RealPromise = Promise
const realFetch = fetch


// So we can resolve them manually and serialize them
export const mockFetch = (url, init) => {
  let resovleTrigger
  let rejectTrigger

  const cbProxy = (resolve, reject) => {
    resovleTrigger = resolve
    rejectTrigger = reject
  }

  const simulation = new RealPromise(cbProxy)

  timetable.register({
    name: url,
    payload: init,
    resolve: resovleTrigger,
    reject: rejectTrigger
  })

  return simulation
}


export const useMock = () => useGlobally('fetch', mockFetch)
export const useReal = () => useGlobally('fetch', realFetch)