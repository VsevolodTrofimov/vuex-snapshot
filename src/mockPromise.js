import timetable from './timetable'
import {useGlobally} from './utility'


const RealPromise = Promise


export class MockPromise extends RealPromise {
  /**
   * Named promise that can be resolved manually and properly serialized 
   * and registers it in timetable
   * @param {Function} cb 
   * @param {string} name
   */
  constructor(cb, name='Promise') {
    let resovleTrigger
    let rejectTrigger

    // name-only construction
    if(typeof cb === 'string') {
      name = cb
      cb = () => {}
    }

    const cbProxy = (resolve, reject) => {
      resovleTrigger = resolve
      rejectTrigger = reject
      cb(resolve, reject)
    }

    super(cbProxy)

    this.name = name
    this.resolve = resovleTrigger
    this.reject = rejectTrigger

    timetable.register({
      name,
      promise: this,
      payload: cb,
      resolve: resovleTrigger,
      reject: rejectTrigger
    })
  }
}


export const useMock = () => useGlobally('Promise', MockPromise)
export const useReal = () => useGlobally('Promise', RealPromise)