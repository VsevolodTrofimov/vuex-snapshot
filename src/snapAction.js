import Snapshot from './snapshot'
import {simualteResolutions} from './resolutionUtils'
import timetable from '../src/timetable'

const RealPromise = Promise


export const makeCallSnapper = (snapshot, type, cb) => (name, payload) => {
  snapshot.add(`${type}: ${name}`, payload)
  cb(name, payload)
}


/**
 * @typedef {{name:string, type: ("resolve" | "reject"), payload}} Resolution
 */
/**
 * Takes snapshot of action's evaluation
 * @param {Function} action action to test
 * @param {{state, getters, commit: Function, dispatch: Function, payload}} mocks arguments passed to the action, payload is the second argument
 * @param {[(string | Resolution)]} resolutions
 * @param {{autoResovle: Boolean, snapEnv: Boolean}} options
 * @returns  {(string | Promise<string>)}
 */
export const snapAction = (action, mocks, resolutions, options) => {
  const snapshot = new Snapshot()
  const mockCommit = makeCallSnapper(snapshot, 'COMMIT', mocks.commit)
  const mockDispatch = makeCallSnapper(snapshot, 'DISPATCH', mocks.dispatch)

  if(options.snapEnv) {
    snapshot.add('DATA MOCKS', {
      state: mocks.state,
      getters: mocks.getters
    })
    snapshot.add('ACTION CALL', mocks.payload)
  }

  const actionReturn = action({
    commit: mockCommit,
    dispatch: mockDispatch,
    state: mocks.state,
    getters: mocks.getters
  }, mocks.payload)

  if(typeof actionReturn !== 'undefined' && actionReturn instanceof Promise) {
    // action is async
    return new RealPromise((resolve, reject) => {
      actionReturn
        .then(payload => {
          snapshot.add('ACTION RESOLVED', payload)
          resolve(snapshot.value)
        })
        .catch(payload => {
          snapshot.add('ACTION REJECTED', payload)
          resolve(snapshot.value)
        })
      
      simualteResolutions(resolutions, snapshot, timetable, options)
        .then(() => {
          // this is needed to let action to resolve first
          setTimeout(() => {
            snapshot.add('ACTION DID NOT RESOLVE')
            resolve(snapshot.value)
          }, 0)
        })
        .catch(err => {
          reject({
            err,
            run: snapshot.value
          })
        })

    })
  } else {
    // action is sync
    return snapshot.value
  }
}

export default snapAction