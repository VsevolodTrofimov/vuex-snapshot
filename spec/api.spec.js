import * as promiseLib from '../src/mockPromise'
import * as fetchLib from '../src/mockFetch'

// fancy mocking
const snapActionModule = require('../src/snapAction')
const snapActionCore = snapActionModule.default
const snapAction = jest.fn(snapActionCore)
snapActionModule.default = snapAction

// the ones being tested
import vuexSnapshot from '../index'
import config from '../src/config'


describe('API', () => {
  describe('config', () => {
    it('Has autoResolve option, with default of false', () => {
      expect(vuexSnapshot.config.autoResolve).toBe(false)
    })
    
    it('Has snapEnv option, with default of false', () => {
      expect(vuexSnapshot.config.snapEnv).toBe(false)
    })
    
    it('Has allowManualActionResolution option, with default of false', () => {
      expect(vuexSnapshot.config.allowManualActionResolution).toBe(false)
    })
    
    it('Resets config w/ .resetConfig', () => {
      vuexSnapshot.config.autoResolve = true
      vuexSnapshot.config.snapEnv = true
      vuexSnapshot.config.allowManualActionResolution = false
    
      vuexSnapshot.resetConfig()
      expect(vuexSnapshot.config.snapEnv).toBe(false)
      expect(vuexSnapshot.config.autoResolve).toBe(false)
    })
    
    it('Resets all w/ .reset', () => {
      const realTimetableReset = vuexSnapshot.timetable.reset
      const realConfigReset = config.reset
      const realUseRealPromise = promiseLib.useReal
      const realUseRealFetch = fetchLib.useReal
      
      vuexSnapshot.timetable.reset = jest.fn()
      config.reset = jest.fn()
      promiseLib.useReal = jest.fn()
      fetchLib.useReal = jest.fn()
    
      vuexSnapshot.reset()
    
      expect(vuexSnapshot.timetable.reset).toBeCalled()
      expect(config.reset).toBeCalled()
      expect(promiseLib.useReal).toBeCalled()
      expect(fetchLib.useReal).toBeCalled()
    
      vuexSnapshot.timetable.reset = realTimetableReset
      config.reset = realConfigReset
      promiseLib.useReal = realUseRealPromise
      fetchLib.useReal = realUseRealFetch
    })
  })


  describe('snapAction', () => {
    // mocks are being reconstructed, so we check them manually
    const expectMocksPassed = (recivedMocks, realMocks) => {
      expect(recivedMocks.state).toEqual(mocks.state)
      expect(recivedMocks.getters).toEqual(mocks.getters)
      expect(recivedMocks.payload).toEqual(mocks.payload)
      expect(mocks.commit).toBeCalled()
      expect(mocks.dispatch).toBeCalled()
    }

    let action
    const mocks={}, resolutions=[], options={}

    beforeEach(() => {
      vuexSnapshot.reset()
      action = jest.fn(({commit, dispatch}) => {
        commit('a')
        dispatch('b')
      })
      
      mocks.state = {
        isState: true
      }
      mocks.getters = {
        value: [1, 2, 3]
      }
      mocks.commit = jest.fn()
      mocks.dispatch = jest.fn()
      mocks.payload = {isPayload: true}

      resolutions.length = 0

      Object.assign(options, config.options)
    })

    it('Passes options', () => {
      vuexSnapshot.snapAction(action, mocks, resolutions)
      expect(snapAction.mock.calls[0][3]).toEqual(options)
    })

    it('Ensures commit and dispatch', () => {
      mocks.commit = undefined
      mocks.dispatch = undefined

      vuexSnapshot.snapAction(action, mocks, resolutions)

      const recivedMocks = snapAction.mock.calls[0][1]

      expect(typeof recivedMocks.commit).toBe('function')
      expect(typeof recivedMocks.dispatch).toBe('function')
    })

    it('Works with [action] args', () => {
      vuexSnapshot.snapAction(action)

      expect(snapAction).toBeCalledWith(
        action, expect.any(Object), 
        expect.any(Array), 
        config.options,
        expect.any(vuexSnapshot.Snapshot)
      )
    })

    it('Works with [action, mocks] args', () => {
      vuexSnapshot.snapAction(action, mocks)

      const recivedMocks = snapAction.mock.calls[0][1]

      expectMocksPassed(recivedMocks, mocks)
      expect(snapAction).toBeCalledWith(
        action, expect.any(Object), 
        expect.any(Array), 
        config.options,
        expect.any(vuexSnapshot.Snapshot)
      )
    })

    it('Works with [action, resolutions] args', () => {
      vuexSnapshot.snapAction(action, resolutions)

      expect(snapAction).toBeCalledWith(
        action, expect.any(Object), 
        resolutions, 
        config.options,
        expect.any(vuexSnapshot.Snapshot)
      )
    })

    it('Works with [action, mocks, resolutions] args', () => {
      vuexSnapshot.snapAction(action, mocks ,resolutions)

      const recivedMocks = snapAction.mock.calls[0][1]

      expectMocksPassed(recivedMocks, mocks)
      expect(snapAction).toBeCalledWith(
        action, expect.any(Object), 
        resolutions, 
        config.options,
        expect.any(vuexSnapshot.Snapshot)
      )
    })

    it('Works with [action, mocks, resolutions, snapshot] args', () => {
      const snapshot = new vuexSnapshot.Snapshot()
      vuexSnapshot.snapAction(action, mocks, resolutions, snapshot)

      const recivedMocks = snapAction.mock.calls[0][1]

      expectMocksPassed(recivedMocks, mocks)
      expect(snapAction).toBeCalledWith(
        action, expect.any(Object), 
        resolutions, 
        config.options,
        snapshot
      )
    })
  })


  describe('Snapshots (backwards compability check)', () => {
    beforeEach(vuexSnapshot.reset)

    it('Sync action', () => {
      const action = ({commit, dispatch}) => {
        commit('increment', 10)
        commit('decrement')
        dispatch('something useful', {notThisOne: 'yeah'})
      }

      expect(vuexSnapshot.snapAction(action)).toMatchSnapshot()
    })

    it('Sync action w/ env & payload', () => {
      const action = ({commit, dispatch, state, getters}, payload) => {
        commit('increment', state.incBig)
        commit('decrement', getters.sadness)
        dispatch('something useful', payload)
      }

      const mocks = {
        state: {
          irrelevat: false,
          incBig: 100
        },
        getters: {sadness: Infinity},
        payload: {action}
      }

      vuexSnapshot.config.snapEnv = true

      expect(vuexSnapshot.snapAction(action, mocks)).toMatchSnapshot()
    })

    it('Async simple', done => {
      const action = ({commit}) => {
        commit('smth')
        return Promise.resolve()
      }

      vuexSnapshot.snapAction(action)
        .then(run => {
          expect(run).toMatchSnapshot()
          done()
        })
        .catch(done)
    })

    it('Async inner promise', done => {
      vuexSnapshot.useMockPromise()

      const action = ({commit}) => new Promise(resolve => {
        commit('smth')
        new Promise(() => {}).then(data => {
          commit('use')
          resolve()
        })
      })

      vuexSnapshot.config.autoResolve = true
      vuexSnapshot.snapAction(action)
        .then(run => {
          expect(run).toMatchSnapshot()
          done()
        })
        .catch(done)
    })

    it('Async state or getter promise', done => {
      const action = ({commit, state}) => new Promise(resolve => {
        commit('smth')
        state.preloadPromise.then(data => {
          commit('use', data)
          resolve(data)
        })
      })

      const state = {
        preloadPromise: new vuexSnapshot.MockPromise('preload')
      }

      vuexSnapshot.snapAction(action, {state}, [{name: 'preload', payload: 'a lot of data'}])
        .then(run => {
          expect(run).toMatchSnapshot()
          done()
        })
        .catch(done)
    })

    it('Async fetch', done => {
      vuexSnapshot.useMockFetch()

      const action = ({dispatch}, attempts=0) => new Promise((resolve, reject) => {
        const request = fetch('http://api.example.com/whatever')
        request
          .then(res => res.json())
          .then(data => {
            dispatch('useExample', data)
            resolve()
          })
          .catch(message => {
            dispatch('callExampleAPI', attempts + 1)
            reject()
          })
      })

      const resolutions = [{
        name: 'http://api.example.com/whatever',
        payload: {
          json: () => new vuexSnapshot.MockPromise('json')
        }
      }, {
        name: 'json',
        payload: {
          data: 'yep',
          important: true
        }
      }]

      vuexSnapshot.config.datTest = true
      vuexSnapshot.snapAction(action, resolutions)
        .then(run => {
          expect(run).toMatchSnapshot()
          done()
        })
        .catch(err => {
          done(err)
        })
      vuexSnapshot.config.datTest = false
    })

    it('Async rejected', done => {
      vuexSnapshot.config.allowManualActionResolution = true
      vuexSnapshot.useMockFetch()

      const action = ({dispatch}, attempts=0) => {
        const request = fetch('http://api.example.com/whatever')
        
        request.catch(message => {
          dispatch('callExampleAPI', attempts + 1)
        })

        return request
      }

      const resolutions = [{
        name: 'http://api.example.com/whatever',
        type: 'reject',
        payload: new TypeError('Failed to fetch')
      }]

      vuexSnapshot.snapAction(action, resolutions)
        .then(run => {
          expect(run).toMatchSnapshot()
          done()
        })
        .catch(done)
    })
  })
})