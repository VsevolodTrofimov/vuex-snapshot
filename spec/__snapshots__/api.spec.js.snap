// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Snapshots (backwards compability check) Async fetch 1`] = `
Array [
  Object {
    "message": "RESOLUTION: http://api.example.com/whatever -> resolve",
    "payload": Object {
      "json": [Function],
    },
  },
  Object {
    "message": "RESOLUTION: json -> resolve",
    "payload": Object {
      "data": "yep",
      "important": true,
    },
  },
  Object {
    "message": "DISPATCH: useExample",
    "payload": Object {
      "data": "yep",
      "important": true,
    },
  },
  Object {
    "message": "ACTION RESOLVED",
  },
]
`;

exports[`API Snapshots (backwards compability check) Async inner promise 1`] = `
Array [
  Object {
    "message": "COMMIT: smth",
  },
  Object {
    "message": "RESOLUTION: Promise -> resolve",
  },
  Object {
    "message": "COMMIT: use",
  },
  Object {
    "message": "ACTION RESOLVED",
  },
]
`;

exports[`API Snapshots (backwards compability check) Async rejected 1`] = `
Array [
  Object {
    "message": "RESOLUTION: http://api.example.com/whatever -> reject",
    "payload": [TypeError: Failed to fetch],
  },
  Object {
    "message": "DISPATCH: callExampleAPI",
    "payload": 1,
  },
  Object {
    "message": "ACTION REJECTED",
    "payload": [TypeError: Failed to fetch],
  },
]
`;

exports[`API Snapshots (backwards compability check) Async simple 1`] = `
Array [
  Object {
    "message": "COMMIT: smth",
  },
  Object {
    "message": "ACTION RESOLVED",
  },
]
`;

exports[`API Snapshots (backwards compability check) Async state or getter promise 1`] = `
Array [
  Object {
    "message": "COMMIT: smth",
  },
  Object {
    "message": "RESOLUTION: preload -> resolve",
    "payload": "a lot of data",
  },
  Object {
    "message": "COMMIT: use",
    "payload": "a lot of data",
  },
  Object {
    "message": "ACTION RESOLVED",
    "payload": "a lot of data",
  },
]
`;

exports[`API Snapshots (backwards compability check) Sync action 1`] = `
Array [
  Object {
    "message": "COMMIT: increment",
    "payload": 10,
  },
  Object {
    "message": "COMMIT: decrement",
  },
  Object {
    "message": "DISPATCH: something useful",
    "payload": Object {
      "notThisOne": "yeah",
    },
  },
]
`;

exports[`API Snapshots (backwards compability check) Sync action w/ env & payload 1`] = `
Array [
  Object {
    "message": "DATA MOCKS",
    "payload": Object {
      "getters": Object {
        "sadness": Infinity,
      },
      "state": Object {
        "incBig": 100,
        "irrelevat": false,
      },
    },
  },
  Object {
    "message": "ACTION CALL",
    "payload": Object {
      "action": [Function],
    },
  },
  Object {
    "message": "COMMIT: increment",
    "payload": 100,
  },
  Object {
    "message": "COMMIT: decrement",
    "payload": Infinity,
  },
  Object {
    "message": "DISPATCH: something useful",
    "payload": Object {
      "action": [Function],
    },
  },
]
`;
